### ISSUE TITLE
[FEA] Add segmented scan feature

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#1365]
**Is your feature request related to a problem? Please describe.**
I wish libcudf would provide a _segmented_ scan feature in addition to the normal scan operation.

**Describe the solution you'd like**
libcudf should add a `cudf::segmented_scan` API that performs a segmented scan operation. 

**Additional context**
This would be a pretty straightforward application of `thrust::scan_by_key` https://thrust.github.io/doc/group__segmentedprefixsums.html. The main question is how we'd like to accept the `keys` parameter. 

One idea is that we could accept a list of offsets that demarcate the beginning of each new segment. For example, the output from a groupby w/o agg. From this, we would have to build the `keys` list parameter for `thrust::scan_by_key`.

Another idea is to require the user to build the `keys` list themselves and we could just use that directly. One question is what type to use for the `keys`? Perhaps `GDF_BOOL`?

**Comment by harrism** [comment]
Is this needed externally to libcudf? If not, it can be done internally with thrust / cub.

**Comment by jrhemstad** [comment]
> Is this needed externally to libcudf? If not, it can be done internally with thrust / cub.

Yeah, it's needed as a public-facing API like `cudf::segmented_scan` such that you could do something like I described here: https://github.com/rapidsai/cudf/issues/1298#issuecomment-480966255

**Comment by harrism** [comment]
I don't understand why you can't just call `thrust::scan_by_key` inside of the groupby `cumsum`.  Or do you need to segmented scan multiple columns and want to abstract it for that reason?  It's still not clear to me why this needs to add to our public API.

**Comment by jrhemstad** [comment]
> I don't understand why you can't just call `thrust::scan_by_key` inside of the groupby `cumsum`. Or do you need to segmented scan multiple columns and want to abstract it for that reason? It's still not clear to me why this needs to add to our public API.

It _could_ be internal, but I don't see any reason to _not_ make it external. 

I can't predict what future need there may be for it to be external, but it's a useful function to have and easy to make into an external API. So why not do so? 

Furthermore, I don't like the idea of adding `cumsum` as an option for the `libcudf` groupby. As we continue to add more complex groupby aggregations, maintaining all of the necessary logic and codepaths to support the various options will become a quagmire to handle in C++. I'd prefer to provide the necessary primitives to Python and let it handle the necessary orchestration logic. Python is much better for that kind of thing.

**Comment by harrism** [comment]
Reasons not to make `segmented_scan` an external API.

1. libcudf is a dataframe library, not a data parallel algorithms library. There are plenty of existing data parallel algorithms library.
2. It increases our API and therefore our testing and maintenance burden
3. We can't build and support features for unpredicted future needs, even when they are useful functions.  No matter how easy, there is no such thing as free ~software~ external APIs.

I don't agree in general that we can rely on Python to orchestrate complex algorithms. libcudf has non-Python clients. We would like important features like groupby to be complete on the C++ side when possible. We can certainly leave much of the quagmire of python/Pandas-specific options to the Python side, but I think the aggregations themselves are squarely in libcudf territory.

I think we should revisit this if and when the need arises.

**Comment by jlowe** [comment]
Speaking of segmented operations and aggregations, what is the plan for supporting aggregations on a sorted grouping approach like that in #982?  Will libcudf provide segmented APIs for all the potential agg functions (sum/max/min/avg/quantiles/etc), or is it up to the app to leverage other algorithm libs like thrust/cub/etc to implement these aggregations on the segments?

**Comment by jrhemstad** [comment]
> Will libcudf provide segmented APIs for all the potential agg functions (sum/max/min/avg/quantiles/etc)

This is the approach I would like to take, which is why I'm suggesting adding a segmented scan function. The follow on would be to add segmented reductions as well. 

> is it up to the app to leverage other algorithm libs like thrust/cub/etc to implement these aggregations on the segments

No, we definitely won't force users to do this.

**Comment by jrhemstad** [comment]
@harrism if we don't provide segmented scan/reduction operations, then the `groupby_without_aggregations` function in https://github.com/rapidsai/cudf/pull/982 is useless to C++ users. 

It organizes the unique groups together and tells you the offset into each group, but there's nothing you can actually do in libcudf with the offset information if you don't have segmented scan/reduction operations.

**Comment by felipeblazing** [comment]
> Furthermore, I don't like the idea of adding cumsum as an option for the libcudf groupby. As we continue to add more complex groupby aggregations, maintaining all of the necessary logic and codepaths to support the various options will become a quagmire to handle in C++. I'd prefer to provide the necessary primitives to Python and let it handle the necessary orchestration logic. Python is much better for that kind of thing than C++ is.

This sort of scares me because we use c++. I would really love for the python side to be as thin as possible to ensure that we can do the same things in c++.

When you talk about cummulative sum this way are you assuming that order matters when going into execution? So for example in thrust

```#include <thrust/scan.h>
#include <thrust/execution_policy.h>
...
int data[10] = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
int keys[10] = {0, 0, 0, 1, 1, 2, 3, 3, 3, 3};
thrust::inclusive_scan_by_key(thrust::host, keys, keys + 10, vals, vals); // in-place scan
// data is now {1, 2, 3, 1, 2, 1, 1, 2, 3, 4};
```

First for this to work we need to order the data according to the keys. This ordering doesn't have to be deterministic and certainly doesn't have to act in a way that the user expects.

In databases windowed functions are used to do things like cummalitive sums since they have clauses that allow you to specify according to which column are they to be ordered.

**Comment by kkraus14** [comment]
> This sort of scares me because we use c++. I would really love for the python side to be as thin as possible to ensure that we can do the same things in c++.

I 100% agree. As we look at other possible language bindings in the future they'd also need to recreate all of the functionality that's implemented in Python. The only things that should be implemented in Python should be Python specific integrations, and Python UDFs.

**Comment by felipeblazing** [comment]
@jrhemstad wouldnt it make more sense to deal with this using the windowed functions

**Comment by jrhemstad** [comment]
> This ordering doesn't have to be deterministic and certainly doesn't have to act in a way that the user expects.

Great point about the ordering. I've asked what the expectation would be in https://github.com/rapidsai/cudf/issues/1298 

> @jrhemstad wouldnt it make more sense to deal with this using the windowed functions

I don't really understand windowed functions enough to know how you could emulate the ask for a groupby quantile/cumsum with a windowed function.

**Comment by jrhemstad** [comment]
For the purposes of further discussion, I wrote up the definitions of the terminology I'm going to use to talk about the different kinds of groupby aggregations. These may not matter to data science end-users, but it matters a lot when implementing these things.

https://gist.github.com/jrhemstad/f6bb5d4c4598a6813493e1c37f01ea02

**Comment by jrhemstad** [comment]
Okay, so my primary anxiety is about trying to mix support for all 3 types of aggregations in a single `cudf::groupby` C++ API.

So what I propose is providing 3 separate APIs:
`cudf::distributive_groupby`
`cudf::algebraic_groupby`
`cudf::holistic_groupby`

(The names aren't set in stone, but you get the idea: different APIs for different classes of groupby.)

In this way, it's clear to the libcudf user that in order to get a `cumsum` groupby requires a vastly different code path than a `min` groupby.

**Comment by harrism** [comment]
Back to my original argument. What I'm suggesting is that if the problem we're solving is supporting aggregations on groupby groups, then we provide aggregations on groupby groups, not a segmented scan in the *public* libcudf API. Underneath, the aggregations may be implemented using a segmented scan (probably Thrust or CUB), but the public API that users access should use familiar terminology (e.g. along the lines of `groupby.cumsum` or `groupby.aggregate`).

**Comment by jrhemstad** [comment]
> Back to my original argument. What I'm suggesting is that if the problem we're solving is supporting aggregations on groupby groups, then we provide aggregations on groupby groups, not a segmented scan in the _public_ libcudf API. Underneath, the aggregations may be implemented using a segmented scan (probably Thrust or CUB), but the public API that users access should use familiar terminology (e.g. along the lines of `groupby.cumsum` or `groupby.aggregate`).

Agreed, so long as we're fine with having independent C++ APIs for the different classes of groupby (as I suggested above).

@harrism and @kkraus14 have agreed to this elsewhere, so I'm going to close the issue for now.

**Comment by jrhemstad** [comment]
> Back to my original argument. What I'm suggesting is that if the problem we're solving is supporting aggregations on groupby groups, then we provide aggregations on groupby groups, not a segmented scan in the _public_ libcudf API. Underneath, the aggregations may be implemented using a segmented scan (probably Thrust or CUB), but the public API that users access should use familiar terminology (e.g. along the lines of `groupby.cumsum` or `groupby.aggregate`).

Agreed, so long as we're fine with having independent C++ APIs for the different classes of groupby (as I suggested above).

@harrism and @kkraus14 have agreed to this elsewhere, so I'm going to close the issue for now.
